### ISSUE TITLE
Ultrasonic "network error" for specific user on requests like "Recently added"

### ISSUE THREAD

**Issue by palto42** [ampache/ampache#1658]
# Environment

 * Ampache version: 8.3.5-develop
 * Web server + version: Nginx 1.10.3-1+deb9u1
 * PHP7
 * Server operating system: Raspbian Stretch
 * Client operating system: Android 7.0

 * Client type:
   - Andorid Ultrasonic
 
# Settings

# Description

Recently I started to get issues with the Ultrasonic client on Android that I only get an error message   on the requests for "Recently Added", "Recently Played", "Most Played", "Starred".
The error message is not helpful since it suggests a network error which is wrong because other functions with Ultrasonic client work fine and I can play music. The error message shown is:
"Error: A network error occurred. Please check the server address or try again later."

On the web interface all this works fine and I don't face any issues. I tried the catalog check, clean and update options, but no change. So I installed a fresh Ampache in parallel with new database, accessing the same music files on the server. With this everything works fine so far, no errors with Ultrasonic client. 

Finally I has the idea to just create a new user on the old Ampache setup and interestingly this also works fine with Ultrasonic. So it seems to be related to the user account and not the music catalog. I only wonder why this problem only happens with Ultrasonic (Subsonic API) and not on the Web interface.

Any idea how I can fix this issue (user database?)? I don't want to lose all my user ratings etc.

# Logs
Couldn't find anything related in the logs. I have enabled logging and set the level to 5 (and restarted Nginx & PHP), but if I access my server from Ultrasonic client, nothing is logged. Have I missed something?

**Comment by wagnered** [comment]
@palto42,

The only difference I can see is that the web interface uses "Statistics Day Threshold" which defaults to 7, and the subsonic API uses "Popular Threshold" which defaults to 0. Both are on the admin->interface tab.

It might be possible that the "object_count" table is corrupted. There is a chance that it can be repaired using the instructions in the  [MySQL 5\.7 Reference Manual](https://dev.mysql.com/doc/refman/5.7/en/repair-table.html).

I tried to duplicate your problem by entering weird values for "Popular Threshold", but still could not duplicate your issue.

**Comment by palto42** [comment]
@wagnered ,

Thanks for the quick response. I tried the repair option, but no change. 
I also tried to delete the whole table content (TRUNCATE object_count), but no change. It still shows the error for my original user account and no error for the new test account.

What's strange to me is that this issue is user dependent although the query seems generic, e.g. "Recently Added" provides what was recently added to the catalog which is the same for both users.
Any other user related table which is used in this query and my cause the issue?

**Comment by wagnered** [comment]
@palto42 

There is an undocumented debug level of 6 which records all or most database queries. Perhaps that might reveal something.

The last resort might be to replace the old user id with the new user id in the "object_count" table.

.Otherwise, I've run out of ideas.

**Comment by palto42** [comment]
@wagnered 
I tried debug level 6 and found one extra SQL query in the log file for the old user account which has an issue on Ultrasonic client which is missing with the new test user connecting to the same server via the same Ultrasonic app (just switching user accounts in the app).

The SQL query (03) below is only done for the old user and not for the new user account. This is done each time after the previous album query as in (02). Is this because of some user preference settings and how could it be related to the issue with the Ultrasonic client?

```
(01)[test] (Query) -> SELECT DISTINCT(`album`) as `id`, MIN(`addition_time`) AS `real_atime` FROM `song` LEFT JOIN `catalog` ON `catalog`.`id` = `song`.`catalog` GROUP BY `album` ORDER BY `real_atime` DESC LIMIT 20 null 
(02)[test] (Query) -> SELECT * FROM `album` WHERE `id`='727' null 
(03)[test] (Query) -> SELECT DISTINCT `album`.`id`, `album`.`disk` FROM album LEFT JOIN `song` ON `song`.`album`=`album`.`id` LEFT JOIN `catalog` ON `catalog`.`id` = `song`.`catalog` WHERE `album`.`mbid`='dc3ca19e-68f9-4fbc-9557-b24739fec00c'  ORDER BY `album`.`disk` ASC null 
(04)[test] (Query) -> SELECT COUNT(DISTINCT(`song`.`artist`)) AS `artist_count`, COUNT(`song`.`id`) AS `song_count`, SUM(`song`.`time`) as `total_duration` FROM `song` WHERE `song`.`album` IN (727) GROUP BY `song`.`artist`  null 
(05)[test] (Query) -> SELECT `song`.`catalog` as `catalog_id`,`artist`.`name` AS `artist_name`, `artist`.`prefix` AS `artist_prefix`, `artist`.`id` AS `artist_id` FROM `song` INNER JOIN `artist` ON `artist`.`id`=`song`.`artist` WHERE `song`.`album` IN (727) LIMIT 1 null 
```

**Comment by palto42** [comment]
I tried a bit more, looking for any differences in the logs between the two accounts. But suddenly I get same error on the new test account for the "Recently Added" query on Ultrasonic. It happened first directly after using the "Top rated" query. The "Recently Played" still works for the new account (but still not on the old one.

Since I did nothing else than this queries from the Ultrasonic client (not played songs or anything else) and there are no other active users, I wonder what this now means. Can a query trigger some DB issue?

I will stop for today, running out of ideas...

**Comment by wagnered** [comment]
@palto42 

Query (03) is only seen when the album has a disk number, musicbrainz id, and "Album-group multiple disks" is enabled. So I'm not  sure this is relevant.

As "Top Rated" from the web interface only reads the db,  It convinces me more than ever that the object_count table might be bad. You could try exporting object_count and user_preference and any other user_* table from the old db and import it into the new db. 

I'm also thinking that maybe the drive might have some problems.  If it has SMART capabilities, It might reveal something.

**Comment by palto42** [comment]
@wagnered 
I have checked my drive and the SMART tests show no errors. Since I also run a Nextcloud server on the same disk without any issues and the Amapache web interface works without issue, I don't think that there is a general disk or database issue.

With try and error testing I now found another hint which might help to find the root cause. As mentioned in my last comment, the test account also started to show this strange error. Via browsing the SQL tables I found one entry of this user in the *user_flag* table. After deleting this entry, the problem was gone. As I then liked (Ultrasonic star = Ampache heart) one album, the problem immediately re-appeared. Removin the "like" in the web interface also fixed the issue. 
Then I "liked" an older album and found that this had no impact, so I suspect that it only causes an issue if the liked album is in the search list (in my case the "recently added albums").

I then tried the same "fix" on my regular account by removing the "likes" (heart) from the latest albums and it also fixed the error on Ultrasonic app.

So there seems to be something wrong with the "like" option on subsonic API I guess.

**Comment by wagnered** [comment]
@palto42 

Good catch.  I was able to duplicate your issue and can now troubleshoot it.
I just added several songs as favorite and the Ultrasonic error showed up.

Thanks....

**Comment by wagnered** [comment]
@palto42
I just pushed the changes needed to cure this issue with commit # cb24ef1b2.

Again, thanks for the help.

**Comment by palto42** [comment]
Closed because the fix is included in new version 3.8.5
